### DISCUSSION TITLE
Simplifier l'API de publicodes

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#64] [Ideas]
Pour utiliser publicodes il faut instancier l'objet moteur en JavaScript avec une base de règle puis on a accès aux méthodes suivantes :

```js
engine.setSituation({ruleName: value})
engine.evaluate('expression')

engine.shallowCopy() // non documenté publiquement
```

Je pense que l'on pourrait simplifier l'API et avoir

```js
engine.set('rule', value)
engine.get('expression')
```

`get` me semble un meilleur nom que `evaluate`, car ce que l'on souhaite c'est récupérer la valeur d'une règle ou d'une expression et pas forcément l'évaluer (d'ailleurs si la valeur est déjà en cache `evaluate` ne fait pas d'évaluation et retourne directement le résultat). Le fait que l'on fait de l'évaluation paresseuse qui la repousse jusqu'à la lecture d'une valeur peut être bien explicité dans la documentation, sans qu'il soit nécessaire de le rappeler dans le nom de la méthode de lecture.

Pour `set`, avoir un appel règle par règle permet de simplifier une logique de cache incrémental #54 (autrement il faut faire un diff entre deux objets situations pour savoir quelles règles ont été modifiées dans la nouvelle situation). Avoir un objet “situation” me semble aussi souvent moins direct dans une UI formulaire que d'avoir la valeur de chaque champ (sur mon-entreprise, on passe par Redux pour reconstituer l'objet situation que l'on donne au moteur).

Comme discuté dans #54, on pourrait avoir une API immuable et supprimer la méthode `shallowCopy`. `engine.set` retournerait un “nouveau moteur” sans modifier l'instance initiale. Ça implique plus de modifications du code qu'un simple renommage des méthodes, mais ça me semble un design d'API souhaitable.

```js

import publicodes from 'publicodes'
import modeleSocial from 'modele-social'

publicodes(modeleSocial)
  .set('contrat salarié . rémunération . nette', '2500 €/mois')
  .set('contrat salarié . temps de travail', '39 €/semaine')
  .get('contrat salarié . rémunération . brute')
```

Il y a sans doute un peu de travail pour que le cache et le ramasse miette fonctionnent correctement avec ce type d'API fonctionnelle. À voir aussi si ça simplifie vraiment notre utilisation côté mon-entreprise (interfaçage avec React, simulateurs, comparaisons de plusieurs situations synchronisées, etc.).

**Comment by johangirod** [comment]
Une remarque principale, je ne suis pas sur de suivre sur `get` vs `evaluate` 
-  `get` est très proche de `getRule`, et les deux ne sont pas tout à fait identique sémantiquement.
-  Par ailleurs, c'est assez générique, et cache le fait que le calcul peut-être couteux. 
- Enfin dans une logique getter / setter, on s'attends à avoir une symétrie des types (entre ce qui est ` set` et ce que l'on `get`), ce qui n'est pas le cas ici (une `PublicodeExpression | ASTNode` dans `set`,  et un `EvaluatedNode & ASTNode` dans `get`). 

Je n'ai pas très bien saisi pourquoi `evaluate` ne convenait pas ? On demande au moteur de retourner le noeud avec son evaluation. Que celle-ci soit déjà en cache ou non est un détail d'implémentation, non ?

Sinon sur le principe du moteur stateless, ça me parait une excellente idée, même si il y a de gros enjeux au niveau du ramasse miette et du partage du cache. Peut-on imaginer un cache partagé entre toutes les instances ? De telle sorte à ce qu'une évaluation sur une partie de l'arbre non modifiée par `set` soit répercutée dans toutes les autres ?
